vowpalwabbit/ftrl.cc

```diff
@@ -401,33 +401,57 @@ base_learner* ftrl_setup(VW::setup_base_i& stack_builder)
-      .add(make_option("coin", coin).keep().help("Coin betting optimizer"))
-      .add(make_option("pistol", pistol).keep().help("PiSTOL: Parameter-free STOchastic Learning"))
+  bool ftrl_option_no_not_use = false;
+  bool pistol_no_not_use = false;
+  bool coin_no_not_use = false;
+
+  option_group_definition ftrl_options("Follow the Regularized Leader - FTRL");
+  ftrl_options
+      .add(make_option("ftrl", ftrl_option_no_not_use)
+               .necessary()
+               .keep()
+               .help("FTRL: Follow the Proximal Regularized Leader"))
       .add(make_option("ftrl_alpha", b->ftrl_alpha).help("Learning rate for FTRL optimization"))
       .add(make_option("ftrl_beta", b->ftrl_beta).help("Learning rate for FTRL optimization"));
 
-  options.add_and_parse(new_options);
+  option_group_definition pistol_options("Follow the Regularized Leader - Pistol");
+  pistol_options
+      .add(make_option("pistol", pistol_no_not_use)
+               .necessary()
+               .keep()
+               .help("PiSTOL: Parameter-free STOchastic Learning"))
+      .add(make_option("ftrl_alpha", b->ftrl_alpha).help("Learning rate for FTRL optimization"))
+      .add(make_option("ftrl_beta", b->ftrl_beta).help("Learning rate for FTRL optimization"));
+
+  option_group_definition coin_options("Follow the Regularized Leader - Coin");
+  coin_options.add(make_option("coin", coin_no_not_use).necessary().keep().help("Coin betting optimizer"))
+      .add(make_option("ftrl_alpha", b->ftrl_alpha).help("Learning rate for FTRL optimization"))
+      .add(make_option("ftrl_beta", b->ftrl_beta).help("Learning rate for FTRL optimization"));
+
+  const auto ftrl_enabled = options.add_parse_and_check_necessary(ftrl_options);
+  const auto pistol_enabled = options.add_parse_and_check_necessary(pistol_options);
+  const auto coin_enabled = options.add_parse_and_check_necessary(coin_options);
+
+  if (!ftrl_enabled && !pistol_enabled && !coin_enabled) { return nullptr; }
+  size_t count = 0;
+  count += ftrl_enabled ? 1 : 0;
+  count += pistol_enabled ? 1 : 0;
+  count += coin_enabled ? 1 : 0;
 
-  if (!ftrl_option && !pistol && !coin) { return nullptr; }
+  if (count != 1) { THROW("You can only use one of 'ftrl', 'pistol', or 'coin' at a time."); }
 
   // Defaults that are specific to the mode that was chosen.
-  if (ftrl_option)
+  if (ftrl_enabled)
   {
     b->ftrl_alpha = options.was_supplied("ftrl_alpha") ? b->ftrl_alpha : 0.005f;
     b->ftrl_beta = options.was_supplied("ftrl_beta") ? b->ftrl_beta : 0.1f;
   }
-  else if (pistol)
+  else if (pistol_enabled)
   {
     b->ftrl_alpha = options.was_supplied("ftrl_alpha") ? b->ftrl_alpha : 1.0f;
     b->ftrl_beta = options.was_supplied("ftrl_beta") ? b->ftrl_beta : 0.5f;
   }
-  else if (coin)
+  else if (coin_enabled)
   {
     b->ftrl_alpha = options.was_supplied("ftrl_alpha") ? b->ftrl_alpha : 4.0f;
     b->ftrl_beta = options.was_supplied("ftrl_beta") ? b->ftrl_beta : 1.0f;
```

Comment:
nit: Can we combine the `if...else/ifs` in lines 444-458 with those on lines 469-500?